subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/ArtifactVariantSelector.java

```diff
@@ -20,25 +20,44 @@
 import org.gradle.api.internal.artifacts.ivyservice.resolveengine.artifact.ResolvedVariant;
 import org.gradle.api.internal.artifacts.ivyservice.resolveengine.artifact.ResolvedVariantSet;
 import org.gradle.api.internal.attributes.ImmutableAttributes;
+import org.gradle.internal.component.SelectionFailureHandler;
+import org.gradle.internal.component.model.GraphVariantSelector;
+
+/**
+ * Selects artifacts from a set of resolved variants, this can but does not necessarily require an additional
+ * round of attribute matching to select a variant containing artifacts.
+ *
+ * This class is intentionally named similarly to {@link GraphVariantSelector}, as it has a
+ * similar purpose.  An instance of {@link SelectionFailureHandler} should be provided
+ * to allow the caller to handle failures in a consistent way - all matching failures should be reported via
+ * calls to that instance.
+ */
+public interface ArtifactVariantSelector {
 
-public interface VariantSelector {
     /**
      * Selects matching artifacts from a given set of candidates.
      *
      * On failure, returns a set that forwards the failure to the {@link org.gradle.api.internal.artifacts.ivyservice.resolveengine.artifact.ArtifactVisitor}.
      */
-    ResolvedArtifactSet select(ResolvedVariantSet candidates, Factory factory);
+    ResolvedArtifactSet select(ResolvedVariantSet candidates, ResolvedArtifactTransformer resolvedArtifactTransformer);
 
     /**
-     * As per {@link #select(ResolvedVariantSet, Factory)} but ignores no matching variants.
+     * As per {@link #select(ResolvedVariantSet, ResolvedArtifactTransformer)} but ignores no matching variants.
      */
-    default ResolvedArtifactSet maybeSelect(ResolvedVariantSet candidates, Factory factory) {
-        return select(candidates, factory);
+    default ResolvedArtifactSet maybeSelect(ResolvedVariantSet candidates, ResolvedArtifactTransformer resolvedArtifactTransformer) {
+        return select(candidates, resolvedArtifactTransformer);
     }
 
     ImmutableAttributes getRequestedAttributes();
 
-    interface Factory {
+    /**
+     * Returns the failure processor which must be used to report failures during variant selection.
+     *
+     * @return the failure processor
+     */
+    SelectionFailureHandler getFailureProcessor();
```

Comment:
Why was this method added? I can't seem to find where it is used

Reply:
`LocalComponentDependencyMetadata#selectVariants` uses it.

Reply:
It calls the method on the `GraphVariantSelector`, not the `ArtifactVariantSelector`

Reply:
OK, this is just leftovers from when I had thought of unifying `GraphVariantSelector` and `ArtifactVariantSelector` into a single type hierarchy.  I'll remove this.